playground/physx-joint-basic.ts

```diff
@@ -0,0 +1,187 @@
+        fixedJoint.connectedCollider = prevCollider;
+      } else {
+        fixedJoint.connectedAnchor = position;
+      }
+      prevCollider = currentCollider;
+    }
+  }
+
+  createChain(new Vector3(6.0, 10.0, 0.0), new Quaternion(), 10, 2.0);
+
+  function createSpring(position: Vector3, rotation: Quaternion) {
+    const currentEntity = addBox(new Vector3(8, 1, 1), position, rotation);
+    const springJoint = currentEntity.addComponent(SpringJoint);
+    springJoint.connectedAnchor = position;
+    springJoint.swingOffset = new Vector3(0, 1, 0);
+    springJoint.maxDistance = 10;
+    springJoint.stiffness = 0.2;
+    springJoint.damping = 1;
+  }
+
+  createSpring(new Vector3(2.0, 10.0, 1.0), new Quaternion());
+
+  function createHinge(position: Vector3, rotation: Quaternion) {
+    const currentEntity = addBox(new Vector3(4.0, 4.0, 0), position, rotation);
+    const hingeJoint = currentEntity.addComponent(HingeJoint);
+    hingeJoint.connectedAnchor = position;
+    hingeJoint.swingOffset = new Vector3(0, 1, 0);
+    hingeJoint.axis = new Vector3(0, 1, 0);
+  }
+
+  createHinge(new Vector3(0, 0, 0), new Quaternion());
+
+  function addSphere(radius: number, position: Vector3, rotation: Quaternion, velocity: Vector3): Entity {
+    const mtl = new BlinnPhongMaterial(engine);
+    mtl.baseColor.set(Math.random(), Math.random(), Math.random(), 1.0);
+    const sphereEntity = rootEntity.createChild();
+    const renderer = sphereEntity.addComponent(MeshRenderer);
+
+    renderer.mesh = PrimitiveMesh.createSphere(engine, radius);
+    renderer.setMaterial(mtl);
+    sphereEntity.transform.position = position;
+    sphereEntity.transform.rotationQuaternion = rotation;
+
+    const physicsSphere = new SphereColliderShape();
+    physicsSphere.radius = radius;
+    const sphereCollider = sphereEntity.addComponent(DynamicCollider);
+    sphereCollider.addShape(physicsSphere);
+    sphereCollider.linearVelocity = velocity;
+    sphereCollider.angularDamping = 0.5;
+
+    return sphereEntity;
+  }
+
+  class ShootScript extends Script {
+    private _dir = new Vector3();
+
+    onUpdate(deltaTime: number) {
+      const transform = this.entity.transform;
+      const dir = this._dir;
+      transform.getWorldForward(this._dir);
+      dir.scale(50);
+
+      if (this.engine.inputManager.isKeyDown(Keys.Space)) {
+        addSphere(0.5, transform.worldPosition, transform.worldRotationQuaternion, dir);
+      }
+    }
+  }
+
+  // init camera
+  const cameraEntity = rootEntity.createChild("camera");
+  cameraEntity.addComponent(Camera);
+  cameraEntity.transform.setPosition(3, 1, 22);
+  cameraEntity.transform.lookAt(new Vector3(3, 1, 0));
+  cameraEntity.addComponent(ShootScript);
+
+  // init direct light
+  const light = rootEntity.createChild("light");
+  light.transform.setPosition(-10, 10, 10);
+  light.transform.lookAt(new Vector3());
+  light.addComponent(DirectLight);
+
+  // Run engine
+  engine.run();
+
+});
```

Comment:
This case can be made more interesting by throwing a small ball from the position of clicking the screen:

1. Screen space generates rays
2. Intersection of ray and camera near cutting plane
3. Throw a small ball from the intersection